core/cas-server-core-services/src/main/java/org/apereo/cas/services/DomainServicesManager.java

```diff
@@ -0,0 +1,197 @@
+package org.apereo.cas.services;
+
+import org.apereo.cas.authentication.principal.Service;
+import org.apereo.cas.support.events.service.CasRegisteredServiceDeletedEvent;
+import org.apereo.cas.support.events.service.CasRegisteredServiceSavedEvent;
+import org.apereo.inspektr.audit.annotation.Audit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationEvent;
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.scheduling.annotation.Scheduled;
+
+import javax.annotation.PostConstruct;
+import java.io.Serializable;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Default implementation of the {@link ServicesManager} interface.
+ *
+ * @author Scott Battaglia
+ * @since 3.1
+ */
+public class DomainServicesManager implements ServicesManager, Serializable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultServicesManager.class);
+    private static final long serialVersionUID = -8581398063126547772L;
+
+    private final ServiceRegistryDao serviceRegistryDao;
+
+    @Autowired
+    private transient ApplicationEventPublisher eventPublisher;
+
+    private Map<Long, RegisteredService> services = new ConcurrentHashMap<>();
+    private Map<String, TreeSet<RegisteredService>> domains = new ConcurrentHashMap<>();
+
+    Pattern domainPattern = Pattern.compile("^https?://([^:/]+)");
```

Comment:
Also try to call RegexUtils to create the pattern, and probably the pattern should be defined as case insensitive?